recirq/toric_code/optimizers.py

```diff
@@ -0,0 +1,452 @@
+
+    A Pauli operation is inserted if it has the `inserted_tag` tag. The goal of this method is to
+    adjust the circuit to undo the functional changes caused by any inserted Paulis by flipping
+    certain measurement results.
+
+    Args:
+        circuit: The circuit to operate on. Should only contain stabilizer operations.
+        inserted_pauli_tag: The tag indicating which operations have been inserted. Should only be
+            applied to Pauli operations.
+
+    Returns:
+        A (modified_circuit, final_pauli_frame) tuple.
+        modified_circuit: A copy of the given circuit where the `inserted_tag` tags have been
+            removed and certain measurements have been inverted.
+        final_pauli_frame: A Pauli string indicating which qubits at the end of the circuit are
+            still bit flipped and/or phase flipped due to the inserted Paulis. Not important for
+            circuits that measure all qubits at the end.
+    """
+    pauli_frame: cirq.PauliString[cirq.Qid] = cirq.PauliString()
+    modified_circuit = cirq.Circuit()
+
+    for old_moment in circuit:
+        new_moment_ops = []
+        for op in old_moment:
+            new_op = op
+            if inserted_pauli_tag in op.tags:
+                # A new Pauli for our Pauli frame.
+                pauli_frame *= op
+                new_op = op.untagged
+                if len(op.tags) > 1:
+                    new_tags = (tag for tag in op.tags if tag != inserted_pauli_tag)
+                    new_op = new_op.with_tags(*new_tags)
+
+            elif isinstance(op.gate, cirq.MeasurementGate):
+                # A measurement that we may need to invert.
+                gate = op.gate
+                inverted_qubits = {
+                    q for q in op.qubits if pauli_frame.get(q) in (cirq.X, cirq.Y)
+                }
+                if inverted_qubits:
+                    new_gate = gate.with_bits_flipped(
+                        *(i for i, q in enumerate(op.qubits) if q in inverted_qubits)
+                    )
+                    new_op = new_gate.on(*op.qubits)
+
+            elif isinstance(op.gate, cirq.ResetChannel):
+                # Reset drops part of the Pauli frame.
+                pauli_frame *= pauli_frame.get(op.qubits[0], cirq.I).on(op.qubits[0])
+
+            else:
+                # This should be a Clifford operation that can transform the Pauli frame.
+                try:
+                    pauli_frame = pauli_frame.pass_operations_over(
+                        [op], after_to_before=True
+                    )
+                except TypeError as ex:
+                    raise NotImplementedError(
+                        f"Don't know how to handle {op!r}."
+                    ) from ex
+
+            new_moment_ops.append(new_op)
+
+        modified_circuit.append(cirq.Moment(new_moment_ops))
+
+    return modified_circuit, pauli_frame / pauli_frame.coefficient
+
+
+def insert_echos_on_active_qubits(
```

Comment:
what's an active qubit? like one that's in the provided `circuit`? Presumably the active qubits are indeed idle during the to-be-inserted echo phase

Reply:
I think active qubits means the ones used in the toric code, but I agree it is a confusing wording.  Changed to idle qubits through this file.